### ISSUE TITLE
How to get the private key and public key from TCert or ECert?

### ISSUE THREAD

**Issue by huhaiqing** [hyperledger-archives/fabric#2240]
Hi,

I know we can get the tcert from memberservice by REST API, such as "GET /registrar/{enrollmentID}/tcert".

But we don't konw how to gets the private key and public key from the returned tcert.
Is there any way to get these two keys. 

We want to use the private key to sign a message and then use the public key to decrypt this message.

Thanks in advance.

Regards
Harry

**Comment by andresgaragiola** [comment]
@huhaiqing you can use [CertificateHandler](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto.go#L132#L133) interface in order to Sign a message using TCert private key. For example to sign a message using a TCert private key you could follow this steps:
- Request a new TCert hanlder using [GetTCertificateHandlerNext](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto.go#L66#L67). An example in [this](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto_test.go#L548#L552) unit tests.
- Use [Sign()](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto.go#L132#L133) function to sign the message with the TCert private key.

If you have a TCert expressed as a DER encoded byte array, you can use   [GetTCertificateHandlerFromDER()](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto.go#L69#L70)  client function to get the CertificateHandler and the client will derive the correct private key, then you can use [Sign()](https://github.com/hyperledger/fabric/blob/master/core/crypto/crypto.go#L132#L133) as was mentioned above. No rest service is available to do what you want to do.

**Comment by smithbk** [comment]
@huhaiqing Hi, can you explain what you are needing to do?  You can use the node.js client SDK to interact with memberservices and chaincode.

**Comment by huhaiqing** [comment]
@smithbk ，Hi, I want to get the TCERT from the REST API, and  use its privare key to sign some message. Then use REST call to invoke a chaincode function, in the chaincode function, I wll use TCERT's publick key to verify the message.

Thanks.

**Comment by huhaiqing** [comment]
@andresgaragiola, For the  CertificateHandler interface, it needs to get a crypto.Client, and this client should be registered to memberservices. In a chaincode function for invoke, however, we don't know how to gets a registered client. Is there any way to get a a registered client ?

Thanks a lot.

**Comment by andresgaragiola** [comment]
@huhaiqing when you invoke a chaincode using REST call "/chaincode" or "/devops/invoke" in both cases a TCert private key is used to sign the whole transaction and this transaction is validated in the peer using the TCert public key. I don't know if this fulfill your requirement, but take in mind that the TCert that you get using REST API is different from the TCert that client will use in the invoke call. 
As I have mentioned above no API is available to do that, probably if you can describe us your use case , we can help you with other solution (for example "attributes"). Any way if you want to get a registered client you can use client.go public functions RegisterClient() and InitClient().

**Comment by JonathanLevi** [comment]
@huhaiqing: Yes, describing what you are trying to do (the use-case), might help to better suggest/advise as to how to best address it.

If easier, you can also use [Slack](http://hyperledgerproject.slack.com/), e.g., there is a dedicated `#membership-services` channel, with similar discussions.
